dotnet tutorial kiota dependency injection (#89)
MicrosoftDocs/openapi-docs: OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md

---
title: Use Kiota with dependency injection in .NET
description: Learn how you can combine Kiota with dependency injection in .NET to build API clients.

> author: svrooij

Reviewer: add your username to this file please https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

Author: Check 😜
Maybe this should be documented somewhere

Reviewer: with #88 we're moving away from that for CI anyway

ms.author: vibiret
ms.topic: tutorial
date: 05/10/2024
---

# Register a Kiota API client in .NET with dependency injection

In this tutorial, you will learn how to register a [Kiota API client](/openapi/kiota/tutorials/dotnet-azure?wt.mc_id=SEC-MVP-5004985) with [dependency injection](/dotnet/core/extensions/dependency-injection-usage?wt.mc_id=SEC-MVP-5004985).

## Required tools

- [.NET SDK 8.0](https://get.dot.net/8)
- [Kiota CLI](/openapi/kiota/install?tabs=bash&wt.mc_id=SEC-MVP-5004985#install-as-net-tool)

## Create a project

Run the following command in the directory you want to create a new project.

```dotnetcli

> dotnet new webapi -o kiota-with-di

Reviewer: add webapi to this file please https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

cd kiota-with-di
dotnet new gitignore
```

## Add dependencies

> Before you can compile and run the generated API client, you will need to make sure the generated source files are part of a project with the required dependencies. Your project must have a reference to the [abstraction package](https://github.com/microsoft/kiota-abstractions-dotnet) and some default implementations.

Reviewer: ```suggestion
Before you can compile and run the generated API client, ensure the generated source files are part of a project with the required dependencies. Your project must reference the [abstraction package](https://github.com/microsoft/kiota-abstractions-dotnet) and default implementations.
```

      properties:
        id:
          type: integer
          description: The release ID
        tag_name:
          type: string
          description: The release tag name
        name:
          type: string
          description: The release name
        body:
          type: string
          description: The release body
        created_at:
          type: string
          format: date-time
          description: The release creation date
        published_at:
          type: string
          format: date-time
          description: The release publication date
    github.error:
      type: object
      properties:
        message:
          type: string
          description: The error message
        documentation_url:
          type: string
          description: The error documentation URL
```

You can then use the [Kiota](/openapi/kiota/install?tabs=bash&wt.mc_id=SEC-MVP-5004985#install-as-net-tool) command line tool to generate the API client classes. We are using the [deserializer](/openapi/kiota/using#--deserializer---ds) and [serializer](/openapi/kiota/using#--serializer---s) options to specify the serialization format, to only use JSON serialization.

```bash

> kiota generate -l csharp -d github-releases.yml -c GitHubClient -n GitHub.ApiClient -o ./GitHub --deserializer Microsoft.Kiota.Serialization.Json.JsonParseNodeFactory --serializer Microsoft.Kiota.Serialization.Json.JsonSerializationWriterFactory

Reviewer: I'd leave the defaults for simplicity and remove the serializer and deserializer arguments. Same for the dependencies earlier.
Other documentation pages explore/will explore the serialization mechanisms and the dependencies, and the choices they can make.

```

## Create extension methods

Create a new file named _KiotaServiceCollectionExtensions.cs_ in the root of your project and add the following code.

```csharp
using Microsoft.Kiota.Http.HttpClientLibrary.Middleware;
using Microsoft.Kiota.Http.HttpClientLibrary.Middleware.Options;

/// <summary>
/// Service collection extensions for Kiota handlers.
/// </summary>
public static class KiotaServiceCollectionExtensions
{
    /// <summary>
    /// Adds the Kiota handlers to the service collection.
    /// </summary>

>     /// <param name="services"><see cref="IServiceCollection"/> to add the services to</param>

Reviewer: add cref to this file https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

    /// <returns><see cref="IServiceCollection"/> as per convention</returns>
    /// <remarks>The handlers are added to the http client by the <see cref="AttachKiotaHandlers(IHttpClientBuilder)"/> call, which requires them to be pre-registered in DI</remarks>
    public static IServiceCollection AddKiotaHandlers(this IServiceCollection services)
    {

>         services.AddTransient<UriReplacementHandler<UriReplacementHandlerOption>>();

Reviewer: out of curiosity here, why not simply call that and feed the array in the services?
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L50

Author: This call is initializing them, and I just need the the types to register them in DI.
Or would you say that doing a 

```csharp

var handlers = KiotaClientFactory.CreateDefaultHandlers()
foreach(var handler in handlers) {
  services.AddTransient(typeof(handler));
}
```

is better? Just trying to see what looks better in coding structures that are commonly used.

Reviewer: The challenge with explicitly enumerating the type is that if tomorrow we add a the middleware handler (let's say something that polls automatically long running operations for example) to the stack, people wouldn't get it automatically.
That's why if we can, it'd be better if it could be more dynamic on that part, even if it means a little more complex code.
does that make sense?

Author: I've made them dynamically, but left the old code. I really don't like this new way because it looks really ugly. And it's creating those handlers during startup twice, just to get the type of them.

Reviewer: assuming this gets implemented, we should be able to clean it up here.
https://github.com/microsoft/kiota-http-dotnet/issues/254

Author: I just noticed that the new way does not work because of this one.
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L55

Reviewer: I believe this was resolved with the update of http. Resolving

        services.AddTransient<RetryHandler>();
        services.AddTransient<RedirectHandler>();
        services.AddTransient<ParametersNameDecodingHandler>();
        services.AddTransient<UserAgentHandler>();
        services.AddTransient<HeadersInspectionHandler>();
        return services;
    }

    /// <summary>
    /// Adds the Kiota handlers to the http client builder.
    /// </summary>
    /// <param name="builder"></param>
    /// <returns></returns>
    /// <remarks>
    /// Requires the handlers to be registered in DI by <see cref="AddKiotaHandlers(IServiceCollection)"/>.
    /// The order in which the handlers are added is important, as it defines the order in which they will be executed.
    /// <see href="https://github.com/microsoft/kiota-http-dotnet/blob/c1c295d3b0ebb2182b66d9a6858241117b59b540/src/KiotaClientFactory.cs#L50-L62">KiotaClientFactory.cs</see> for the default order.
    /// </remarks>
    public static IHttpClientBuilder AttachKiotaHandlers(this IHttpClientBuilder builder)
    {

>         builder.AddHttpMessageHandler<UriReplacementHandler<UriReplacementHandlerOption>>();

Reviewer: same here about using the factory?

Author: I literally have no clue how you could use an instance of something in a method that expects just the type in the generic specifier.

Author: See above

```csharp
using GitHub.ApiClient;
using Microsoft.Kiota.Abstractions.Authentication;
using Microsoft.Kiota.Http.HttpClientLibrary;

namespace GitHub;

public class GitHubClientFactory
{
    private readonly IAuthenticationProvider _authenticationProvider;
    private readonly HttpClient _httpClient;

    public GitHubClientFactory(HttpClient httpClient)
    {
        _authenticationProvider = new AnonymousAuthenticationProvider();
        _httpClient = httpClient;
    }

    public GitHubClient GetClient() {
      return new GitHubClient(new HttpClientRequestAdapter(_authenticationProvider, httpClient: _httpClient));
    }
}
```

## Register the API client

Open the _Program.cs_ file and add the following code above the `var app = builder.Build();` line.

```csharp
using GitHub;
...

> // Add Kiota handlers to the DI container

Reviewer: ```suggestion
// Add Kiota handlers to the dependency injection container
```

builder.Services.AddKiotaHandlers();

// Register the factory for the GitHub client
builder.Services.AddHttpClient<GitHubClientFactory>((sp, client) => {
    client.BaseAddress = new Uri("https://api.github.com");
    client.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json");
}).AttachKiotaHandlers();

// Register the GitHub client

> builder.Services.AddTransient(sp=> sp.GetRequiredService<GitHubClientFactory>().GetClient());

Reviewer: ```suggestion
builder.Services.AddTransient(sp => sp.GetRequiredService<GitHubClientFactory>().GetClient());
```

```csharp
app.MapGet("/dotnet/releases", async (GitHub.ApiClient.GitHubClient client, CancellationToken cancellationToken) =>
{
    var releases = await client.Repos["dotnet"]["runtime"].Releases.Latest.GetAsync(cancellationToken: cancellationToken);
    return releases;
})
.WithName("GetDotnetReleases")
.WithOpenApi();
```

## Run the application

Run the following command in your project directory to start the application.

```dotnetcli
dotnet run
```

You can now open a browser and navigate to `http://localhost:{port}/dotnet/releases` to see the latest release of `dotnet/runtime`. Alternatively, you can also go to `http://localhost:{port}/swagger` to see the Swagger UI with this new endpoint.

## Authentication

The `GitHubClientFactory` class is responsible for creating the client, and it is also responsible for creating the `IAuthenticationProvider`. In this sample, we are using the `AnonymousAuthenticationProvider`, which is a simple provider that does not add any authentication headers. More details on authentication with Kiota can be found in the [Kiota authentication documentation](/openapi/kiota/authentication?wt.mc_id=SEC-MVP-5004985).

> What to remember is that the `GitHubClientFactory` is used through dependency injection, which means you can have other types injected by the constructor that are needed for the authentication provider.

Reviewer: ```suggestion
The `GitHubClientFactory` is used through dependency injection, which means you can have other types injected by the constructor that are needed for the authentication provider.
```

> In [this example](https://github.com/svrooij/KiotaWithDependencyInjection/blob/d1ee972e8ba06c4f9707b78dfa981f5f57fbf7e6/KiotaWithDependencyInjection/Spotify/SpotifyFactory.cs) there is an `IAccessTokenProvider` injected that is then [used](https://github.com/svrooij/KiotaWithDependencyInjection/blob/d1ee972e8ba06c4f9707b78dfa981f5f57fbf7e6/KiotaWithDependencyInjection/Spotify/SpotifyFactory.cs#L36) for the [BaseBearerTokenAuthenticationProvider](/openapi/kiota/authentication?tabs=csharp&wt.mc_id=SEC-MVP-5004985#base-bearer-token-authentication-provider) that is available in Kiota.

Reviewer: links to "3rd party repositories" will lead to scrutiny. I suggest removing this section

Author: No sample better than me creating a pr for the samples repo?

Reviewer: we could alternatively do that if you feel like adding more samples to the kiota-samples repo

Author: Might create the sample in the sample repo and then create a new pr to update this one.
